### PR TITLE
[Audit Fix] N-13 - Gas optimizations -  refactor tokens approval

### PR DESCRIPTION
**Pull request by haythem96**
# Task: refactor tokens approval

**Merge after #743**

## Description

Fixes ENG-895

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Document update

## How Has This Been Tested

Please describe how to test to verify the changes. Provide instructions so we can reproduce.

## FE Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] Added video recordings if it is a UI change

## User Facing Checklist

- [ ] I fully understand the user problem this PR is solving
- [ ] I know who the target user is for this PR and have a deep understanding of that user
- [ ] I have tried this flow thinking from the pov of the target user for this PR
- [ ] I (or working w someone on team) have scheduled a user test for this PR (if it is a large change)

### CODE REVIEW

File: packages/bull-vault/src/FlashBull.sol [hunk 1]
```diff
@@ -117,6 +117,14 @@ contract FlashBull is UniFlash {
         ethWSqueethPool = IController(IBullStrategy(_bull).powerTokenController()).wPowerPerpPool();
         ethUSDCPool =
             IController(IBullStrategy(_bull).powerTokenController()).ethQuoteCurrencyPool();
+
+        ICrabStrategyV2(IBullStrategy(_bull).crab()).approve(_bull, type(uint256).max);
+        IERC20(IController(IBullStrategy(_bull).powerTokenController()).wPowerPerp()).approve(
```
Comment (alpinechicken) [inline packages/bull-vault/src/FlashBull.sol:122]: can just use wPowerPerp since we set it?

File: packages/bull-vault/src/FlashBull.sol [hunk 2]
```diff
@@ -117,6 +117,14 @@ contract FlashBull is UniFlash {
...
+            _bull, type(uint256).max
+        );
+        IERC20(IController(IBullStrategy(_bull).powerTokenController()).quoteCurrency()).approve(
```
Comment (alpinechicken) [inline packages/bull-vault/src/FlashBull.sol:125]: use usdc?

File: packages/bull-vault/src/FlashBull.sol [hunk 3]
```diff
@@ -117,6 +117,14 @@ contract FlashBull is UniFlash {
         ethWSqueethPool = IController(IBullStrategy(_bull).powerTokenController()).wPowerPerpPool();
         ethUSDCPool =
             IController(IBullStrategy(_bull).powerTokenController()).ethQuoteCurrencyPool();
+
+        ICrabStrategyV2(IBullStrategy(_bull).crab()).approve(_bull, type(uint256).max);
```
Comment (alpinechicken) [inline packages/bull-vault/src/FlashBull.sol:121]: use crab

**Review by alpinechicken — APPROVED** [review]
lgtm - comments on only calling interfaces once in constructor just for readability

File: packages/bull-vault/src/FlashBull.sol (hunk 1 above)
Comment (haythem96) [reply packages/bull-vault/src/FlashBull.sol:122]: `wPowerPerp` is immutable, so we can't use it in the constructor when we set it also

File: packages/bull-vault/src/FlashBull.sol (hunk 2 above)
Comment (haythem96) [reply packages/bull-vault/src/FlashBull.sol:125]: same as wPowerPerp

File: packages/bull-vault/src/FlashBull.sol (hunk 3 above)
Comment (haythem96) [reply packages/bull-vault/src/FlashBull.sol:121]: same as wPowerPerp

**Review by aleone — APPROVED** [review]

**Pull request merged**